js/util/pipeline-fixed.js

```diff
@@ -0,0 +1,350 @@
+    } catch (ex) {
+      console.log("Error in KeyChain.verifyData: " + ex);
+    }
+  }
+  else {
+    this.onVerified(data, originalInterest);
+  }
+};
+
+PipelineFixed.prototype.onVerified = function(data, originalInterest)
+{
+  var currentSegment = 0;
+  try {
+    currentSegment = data.getName().get(-1).toSegment();
+  }
+  catch (ex) {
+    this.reportError(PipelineFixed.ErrorCode.DATA_HAS_NO_SEGMENT,
+                     "Error decoding the name segment number " +
+                     data.getName().get(-1).toEscapedString() + ": " + ex);
+    return;
+  }
+
+  // if segment number of the very first Data packet does not start
+  // with zero, we have to start asking from segment zero
+  if (this.firstSegmentIsReceived === false) {
+    this.firstSegmentIsReceived = true;
+    if (currentSegment === 0) {
+      this.nextSegmentToRequest += 1;
+    }
+  }
+
+  // set finalBlockId
+  if (data.getMetaInfo().getFinalBlockId().getValue().size() > 0) {
+    try {
+      this.finalBlockId = (data.getMetaInfo().getFinalBlockId().toSegment());
+      this.cancelPendingInterestsAboveFinalBlockId();
+    }
+    catch (ex) {
+      this.reportError(PipelineFixed.ErrorCode.DATA_HAS_NO_SEGMENT,
+           "Error decoding the FinalBlockId segment number " +
+            data.getMetaInfo().getFinalBlockId().toEscapedString() +
+            ": " + ex);
+      return;
+    }
+  }
+
+  this.numberOfSatisfiedSegments += 1;
+
+  // Save the content
+  this.contentParts[currentSegment] = data.getContent().buf();
+
+  // Check whether we are finished
+  if (this.numberOfSatisfiedSegments > this.finalBlockId) {
+    // Concatenate to get content.
+    var content = Buffer.concat(this.contentParts);
+    try {
+      this.onComplete(new Blob(content, false));
+    } catch (ex) {
+      console.log("Error in onComplete: " + NdnCommon.getErrorWithStackTrace(ex));
+    }
+
+    return;
+  }
+
+  this.decreaseNumberOfSegmentsOnFly();
+
+  // Fetch the next segments
+  this.fetchNextSegments
+    (originalInterest, data.getName());
+};
+
+PipelineFixed.prototype.onValidationFailed = function(data, reason)
+{
+  this.reportError(PipelineFixed.ErrorCode.SEGMENT_VERIFICATION_FAILED,
+                   "Segment verification failed for " + data.getName().toUri() +
+                   " . Reason: " + reason);
+};
+
+PipelineFixed.prototype.onTimeout = function(interest)
+{
+  this.reportError(PipelineFixed.ErrorCode.INTEREST_TIMEOUT,
+                   "Time out for interest " + interest.getName().toUri());
+};
+
+PipelineFixed.prototype.onNack = function(interest)
+{
+  this.reportError(PipelineFixed.ErrorCode.NACK_RECEIVED,
+                   "Received Nack for interest " + interest.getName().toUri());
+};
+
+PipelineFixed.prototype.increaseNumberOfSegmentsOnFly = function()
+{
+  this.segmentsOnFly += 1;
+};
+
+PipelineFixed.prototype.decreaseNumberOfSegmentsOnFly = function()
+{
+  this.segmentsOnFly = (this.segmentsOnFly <= 0) ? 0
```

Comment:
Simplify as:

    this.segmentsOnFly = Math.max(this.segmentOnFly - 1, 0);

It’s more readable than ?: operator.